README.txt
$Id: README.txt,v 1.1.2.4 2010/07/02 17:08:49 yhahn Exp $

Features 1.x for Drupal 6.x
---------------------------
The features module enables the capture and management of features in Drupal. A
feature is a collection of Drupal entities which taken together satisfy a
certain use-case.

Features provides a UI and API for taking different site building components
from modules with exportables and bundling them together in a single feature
module. A feature module is like any other Drupal module except that it declares
its components (e.g. views, contexts, CCK fields, etc.) in its `.info` file so
that it can be checked, updated, or reverted programmatically.

Examples of features might be:

- A blog
- A pressroom
- An image gallery
- An e-commerce t-shirt store


Installation
------------
Features can be installed like any other Drupal module -- place it in the
modules directory for your site and enable it on the `admin/build/modules` page.
To take full advantage of some of the workflow benefits provided by Features,
you should install [Drush][1].


Basic usage
-----------
Features is geared toward usage by developers and site builders. It
is not intended to be used by the general audience of your Drupal site.
Features provides tools for accomplishing two important tasks:

### Task 1: Export features

You can build features in Drupal by using site building tools that are supported
(see a short list under the *Compatibility* section).

Once you've built and configured functionality on a site, you can export it into
a feature module by using the feature create page at
`admin/build/features/create`.


### Task 2: Manage features

The features module also provides a way to manage features through a more
targeted interface than `admin/build/modules`. The interface at
`admin/build/features` shows you only feature modules, and will also inform you
if any of their components have been overridden. If this is the case, you can
also re-create features to bring the module code up to date with any changes
that have occurred in the database.


Using Features to manage development
------------------------------------
Because Features provides a centralized way to manage exportable components and
write them to code it can be used during development in conjunction with a
version control like SVN or git as a way to manage changes between development,
staging and production sites. An example workflow for a developer using Features
is to:

1. Make configuration changes to a feature on her local development site.
2. Update her local feature codebase using `drush features-update`.
3. Commit those changes using `svn commit`.
4. Roll out her changes to the development site codebase by running `svn update`
  on the server. Other collaborating developers can also get her changes with
  `svn update`.
5. Reverting any configuration on the staging site to match the updated codebase
by running `drush features-revert`.
6. Rinse, repeat.

Features also provides integration with the [Diff][3] module if enabled to show
differences between configuration in the database and that in code. For site
builders interested in using Features for development, enabling the diff module
and reading `API.txt` for more details on the inner workings of Features is
highly recommended.


Drush usage
-----------
Features provides several useful drush commands:

- `drush features`

  List all the available features on your site and their status.

- `drush features-export [feature name] [component list]`

  Write a new feature in code containing the components listed.

- `drush features-update [feature name]`

  Update the code of an existing feature to include any overrides/changes in
  your database (e.g. a new view).

- `drush features-revert [feature name]`

  Revert the components of a feature in your site's database to the state
  described in your feature module's defaults.

- `drush features-diff [feature name]`

  Show a diff between a feature's database components and those in code.
  Requires the Diff module.

Additional commands and options can be found using `drush help`.


Compatibility
-------------
Features provides integration for the following exportables:

- CTools export API implementers (Context, Spaces, Boxes, Strongarm, Page
  Manager)
- ImageCache
- Views
- [Other contributed modules][2]

Features also provides faux-exportable functionality for the following Drupal
core and contrib components:

- CCK fields
- CCK fieldgroups
- Content types
- Input filters
- User roles/permissions
- Custom menus and menu links *
- Taxonomy vocabularies *

* Currently in development.


For developers
--------------
Please read `API.txt` for more information about the concepts and integration
points in the Features module.


Maintainers
-----------
- yhahn (Young Hahn)
- jmiccolis (Jeff Miccolis)


[1]: http://drupal.org/project/drush
[2]: (http://drupal.org/taxonomy/term/11478)


CHANGELOG.txt
// $Id: CHANGELOG.txt,v 1.1.2.12 2010/07/30 19:23:41 yhahn Exp $

features DRUPAL-6--1-0-BETA11
-----------------------------
- #865454: Fix for complex menu links component handling.
- Context: simpler check for API version
- Blank out .features.inc when appropriate when updating existing features.
- Fix for haphazard drush features-revert.
- Adjust features_get_component_map() to allow for easy retrieval of a specific
  component key.

features DRUPAL-6--1-0-BETA10
-----------------------------
- #864742: Ensure CTools plugins.inc is included before calling API functions.
- #864482 by hadsie: Fix for PHP notice.
- Fix Views dependency detection.

features DRUPAL-6--1-0-BETA9
----------------------------
This release introduces several important API and component changes.

### Include file writing

- CTools and Views components are now written to supported API include files,
  e.g. `[myfeature].views_default.inc` and `[myfeatures].[ctools api].inc`.
- Features-supported faux-exportable components are written to distinct include
  files, e.g. `[myfeature].features.[component].inc`.
- Only API hooks and hooks for APIs that do not handle file inclusion should be
  written to `[myfeatures].features.inc` which is loaded by default through the
  feature module file. These include `hook_views_api`, `hook_ctools_api`,
  `hook_node_info`, etc.
- All features using the old include formats are still supported but many
  components will be marked as 'Overridden' until the feature is updated.
- `[myfeature].defaults.inc`, `[myfeature].features.node.inc` and
  `[myfeature].features.views.inc` are deprecated and can be safely removed
  after the feature has been updated.

### Component API changes

- An exportable taxonomy vocabulary component has been introduced which uses
  the `module` column of the vocabulary table as storage for machine names.
- The menu component has been deprecated in favor of more advanced `menu_links`
  and `menu_custom` components. Features using the old `menu` component can be
  migrated to the new components through a feature update.
- The old user permission component has been deprecated in favor of the
  `user_permission` and `user_role` components. Features will no longer
  automatically create roles that do not exist -- they must be exported
  properly using the `user_role` component. Features using the old `user`
  component can be migrated to the new components through a feature update.

- FIx for features implementers using FEATURES_DEFAULTS_CUSTOM.
- #504706 by dmitrig01: Recreate: include any additional files from the original
  feature
- #728004 by gordon, dmitrig01: Allow abitrary modules to alter the pipeline.
- #752144 by dmitrig01: Refactor Features Status Page for Performance
- #841844 by dmitrig01: Implement environment-agnostic message logging.
- #577674 by dmitrig01: Display reasons for Incompatible Feature.
- #838726: Fix possible creation of circular dependencies by cross-feature
  fields.
- #829778 by Josh Waihi: Add features-diff command.
- Fix for reckless iteration.
- Make features component options DOM safe for AHAH handling.
- #717152 by dmitrig01: Ensure dependent features cannot be disabled.
- Strip tags when displaying log messages in CLI.
- #823278 by Grayside: Display warning when feature .module does not include
  .feature.inc.
- #649604: Split out dependencies as a component, proper, and ensure that
  feature dependencies are met.
- #742940: Improve granularity and compatibility of Features export includes.
- #741498: Fix for bad function signature.
- Suppress PHP 5.3 errors/warnings when generating tar output.
- #813960 by nyl auster: Make it possible to revert defaulted features with
  --force
- Massive update to README.txt.
- #843956 by q0rban: Drush commands use deprecated callback names
- #829778 by q0rban: Offer to download and install diff if it is not available.
- #843106 by dmitrig01: Check hook_requirements.
- Updating API.txt
- #843106 by dagmar: Fix for hook_requirements() checking.
- Adding a note about custom code to README.txt
- #789556 by alex_b, yhahn, bcmiller0 and jmiccolis, Support taxonomy
  vocabularies.
- Adding hook help.
- #853738 by fgm: Fix for reckless check against ['default_file']
- #852704 by q0rban: Check if Taxonomy exists before piping.
- #633804 by catch, dereine, yhahn, JohnAlbin: Improve menu support
- Ensure feature component options are escaped.
- #792790 by Dustin Currie: Fix label for components editing.
- Cleanup display of components on create/reexport.
- Fix for PHP notice.
- Fix for export and allow drush fu to affect modules in "non-standard"
  directories.
- Fix PHP notice when description is not set.
- Fix for PHP notice.
- Reduce excessive cache clearing.
- Refactor features_get_default() and introduce features_get_default_map() to
  simplify implementers' code.
- Update to features_test module.
- Remove misleading log messages.
- #830774 by scottrigby: Fix for reckless iteration.
- #751538 by DeFr: Clear caches after batch revert of Views rather than
  individually.
- #678930 by tim.plunkett: Fix Features package tab sorting.
- Die tabs.
- Remove confusing debug output for revert.
- Don't run the bulk of feature generation/export logic unless we're sure the
  feature can be written.
- Check that specified object is loaded before attempting export.
- #693024 by ceardach, hefox, nedjo: Add alter hooks for faux exportables.
- Only write Views API hook if views are actually exported.
- Fix for bad params.
- Removing excessive taxonomy pipe from node component.
- Preserve selected values even if form fails to validate.
- #849646: CTools: fix for integration with page manager list callback.
- #553866: Split user component into user_roles, user_permissions.
- #650292: Limit export options to normal and current feature's components.
- Fixing user permission test.
- Remove special handling for menu component.
- #649298 by q0rban, KarenS, yhahn: Allow node component type revert to disable

theme/features-form.tpl.php
<?php
// $Id: features-form.tpl.php,v 1.1.2.3.2.1 2010/05/07 17:11:11 adrian Exp $
?>
<div class='features-form clear-block'>
  <?php if (count($packages) > 1): ?>
    <div class='features-form-content'>
      <?php foreach ($packages as $class => $package): ?>
        <div class='features-form-package package-<?php print $class?>'><?php print drupal_render($package) ?></div>
      <?php endforeach; ?>
      <?php print drupal_render($form) ?>
    </div>
    <div class='features-form-links'><?php print theme('links', $package_links, array('id' => 'features-form-links')) ?></div>
  <?php elseif (count($packages) == 1): ?>
    <?php print drupal_render($packages) ?>
    <?php print drupal_render_children($form) ?>
  <?php else: ?>
    <?php print t('No features; consider <a href="!url">adding one.</a>', array('!url' => url('admin/structure/features/create'))); ?>
  <?php endif; ?>
</div>


features.js
// $Id: features.js,v 1.1.2.8.2.1 2010/05/07 17:11:11 adrian Exp $
(function ($) {
  Drupal.behaviors.features = {
    attach: function(context, settings) {
      // Features management form package tabs
      $("ul#features-form-links li a:not(.features-processed)").each(function() {
        $(this).addClass('features.processed').click(function() {
          $(".features-package-active").removeClass('features-package-active');
          var panel = $(this).attr('href').split('#')[1];
          $("div.package-" + panel).addClass('features-package-active');
          $(this).addClass('features-package-active');
          return false;
        });
      });

      // Features management form
      $('table.features:not(.processed)').each(function() {
        $(this).addClass('processed');

        // Check the overridden status of each feature
        Drupal.features.checkStatus();

        // Add some nicer row hilighting when checkboxes change values
        $('input', this).bind('change', function() {
          if (!$(this).attr('checked')) {
            $(this).parents('tr').removeClass('enabled').addClass('disabled');
          }
          else {
            $(this).parents('tr').addClass('enabled').removeClass('disabled');
          }
        });
      });

      // Export form component selector
      $('form.features-export-form select.features-select-components:not(.processed)').each(function() {
        $(this)
          .addClass('processed')
          .change(function() {
            var target = $(this).val();
            $('div.features-select').hide();
            $('div.features-select-' + target).show();
            return false;
        });
      });

      // Export form machine-readable JS
      $('.feature-name:not(.processed)').each(function() {
        $('.feature-name')
          .addClass('processed')
          .after(' <small class="feature-project-suffix">&nbsp;</small>');
        if ($('.feature-project').val() === $('.feature-name').val().toLowerCase().replace(/[^a-z0-9]+/g, '_').replace(/_+/g, '_') || $('.feature-project').val() === '') {
          $('.feature-project').parents('.form-item').hide();
          $('.feature-name').keyup(function() {
            var machine = $(this).val().toLowerCase().replace(/[^a-z0-9]+/g, '_').replace(/_+/g, '_');
            if (machine !== '_' && machine !== '') {
              $('.feature-project').val(machine);
              $('.feature-project-suffix').empty().append(' Machine name: ' + machine + ' [').append($('<a href="#">'+ Drupal.t('Edit') +'</a>').click(function() {
                $('.feature-project').parents('.form-item').show();
                $('.feature-project-suffix').hide();
                $('.feature-name').unbind('keyup');
                return false;
              })).append(']');
            }
            else {
              $('.feature-project').val(machine);
              $('.feature-project-suffix').text('');
            }
          });
          $('.feature-name').keyup();
        }
      });


    }

  }


  Drupal.features = {
    'checkStatus': function() {
      $('table.features tbody tr').not('.processed').filter(':first').each(function() {
        var elem = $(this);
        $(elem).addClass('processed');
        var uri = $(this).find('a.admin-check').attr('href');
        if (uri) {
          $.get(uri, [], function(data) {
            $(elem).find('.admin-loading').hide();
            switch (data.storage) {
              case 3:
                $(elem).find('.admin-rebuilding').show();
                break;
              case 2:
                $(elem).find('.admin-needs-review').show();
                break;
              case 1:
                $(elem).find('.admin-overridden').show();
                break;
              default:
                $(elem).find('.admin-default').show();
                break;
            }
            Drupal.features.checkStatus();
          }, 'json');
        }
        else {
            Drupal.features.checkStatus();
          }
      });
    }
  };


})(jQuery);


